--- usr/lib/mailman/bin/newlist.orig	2011-06-04 16:03:31.000000000 +0200
+++ usr/lib/mailman/bin/newlist	2011-06-04 16:06:45.000000000 +0200
@@ -164,7 +164,7 @@
 
     if '@' in listname:
         # note that --urlhost and --emailhost have precedence
-        listname, domain = listname.split('@', 1)
+        firstname, domain = listname.split('@', 1)
         urlhost = urlhost or domain
         emailhost = emailhost or mm_cfg.VIRTUAL_HOSTS.get(domain, domain)
 
--- usr/lib/mailman/bin/rmlist.orig	2011-06-04 16:03:35.000000000 +0200
+++ usr/lib/mailman/bin/rmlist	2011-06-04 16:06:45.000000000 +0200
@@ -92,6 +92,11 @@
         usage(1)
     listname = args[0].lower().strip()
 
+    if '@' in listname:
+        # note that --urlhost and --emailhost have precedence
+        firstname, domain = listname.split('@', 1)
+        listname = '%s-%s' % ( firstname, domain )
+
     removeArchives = False
     for opt, arg in opts:
         if opt in ('-a', '--archives'):
