--- cgi-wrapper.c.orig	2015-02-18 11:15:42.000000000 +0100
+++ cgi-wrapper.c	2015-02-18 11:16:25.865233254 +0100
@@ -42,8 +42,10 @@
         char* fake_argv[3];
 
         running_as_cgi = 1;
+/* AlternC PATCH for apache mpm itk compliance
         if (getgid() >= 100 && getgid() != 65534)
                 check_caller(logident, parentgroup);
+*/
 
         /* For these CGI programs, we can ignore argc and argv since they
          * don't contain anything useful.  `script' will always be the driver
